docs/java-rest/high-level/migration.asciidoc

```diff
@@ -0,0 +1,137 @@
+[[java-rest-high-level-migration]]
+== Migration Guide
+
+This section describes how to migrate existing code from the `TransportClient`
+to the new Java High Level REST Client released with the version 5.6.0
+of Elasticsearch.
+
+=== Motivations around a new Java client
+
+The existing `TransportClient` has been part of Elasticsearch since https://github
+.com/elastic/elasticsearch/blob/b3337c312765e51cec7bde5883bbc0a08f56fb65/modules/elasticsearch/src/main/java/org/elasticsearch/client
+/transport/TransportClient.java[its very first commit]. Although it covers the whole set of Elasticsearch APIs,
+it always has had various flaws:
+
+- it is tightly coupled with Elasticsearch and uses Elasticsearch's internal implementations
+- it uses a home grown binary serialization to transport objects over the wire
+
+These two aspects forces any Java application that communicates with an Elasticsearch cluster to use
+a `TransportClient` in a version compatible with the remote cluster's version. That's acceptable most of
+the time but Elasticsearch moves fast and new versions are released every month, forcing Java applications
+to use a newer version of the `TransportClient` just because the remote cluster has been upgraded.
+
+We decided to tackle these issues and we released a low-level REST client in 2016. This low-level REST client
+is based on the well-known Apache HTTP Client and it allows to communicate with an Elasticsearch cluster in any
+version using HTTP. On top of it, we released the high-level REST client. It is based on the low-level client
+and it provides a <<java-rest-high-supported-apis,subset of Elasticsearch's APIs>> that uses Elasticsearch core
+Java classes, like the `TransportClient` does, in order to perform requests and to parse responses.
```

Comment:
I don't know - I added this to add some context around why it's good to migrate, since it's easy to miss a blog post. But I can remove this whole section if you really feel it should not be here.